wgsl/index.bs

```diff
@@ -4557,11 +4555,291 @@ TODO: *Stub*: Expression evaluation
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+<div class="note">Note: apart from two special nodes True and False, all nodes can be understood as having one of the following meanings:
+        - The control flow at a specific program point must be uniform
+        - The value of a specific expression must be uniform
+        - A specific variable must always hold a uniform value
+
+        An edge can be understood as an implication from the statement corresponding to its source node to the statement corresponding to its target node.
+        In particular, to express that something must always be uniform (e.g. the control flow at the call site of a derivative), we simply add an edge from True to the corresponding node.
+        Reciprocally, to express that we cannot ensure the uniformity of something (e.g. a variable which holds the thread id), we simply add an edge from the corresponding node to False
+</div>
+
+In more detail, here is the algorithm for analyzing a function:
+
+* Create some nodes called "True", "False", "CF_start", "CF_return" and if the function is non-void a node called "Value_return"
+* Create one node for each parameter of the function which we'll call "arg_i"
+* Walk over the AST of the function, adding nodes and edges to the graph following the rules of the next section.
+* Look at the *Behaviors* of the function body, using the rules for statements in the next section
+    * If it has a behavior which is neither Nothing nor Discard, then reject the program
+    * Else remember it for later.
+* Look at which nodes are reachable from "True"
+    * If this set includes the node "False", then reject the program.
+    * Remember for later whether this set includes "CF_start" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all nodes that have been visited
+* Look at which nodes are reachable from "CF_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all nodes that have been visited
+* If "Value_return" exists, look at which nodes are reachable from it
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+
+Note: Anything which is not explicitly remembered, including the entire graph can be destroyed at this point. The little information that we do need to remember to analyze the caller is linear in the number of parameters of the function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+We achieve this by assigning to each statement a set of *behaviors* out of {Return, Break, Continue, Discard, Fallthrough, Nothing}.
```

Comment:
I think I asked this but didn't see answer. Why is `Nothing` a behavior? Shouldn't `{}` have this semantics?

Reply:
Nothing is a behavior to differentiate between the following programs:
```
return;
```
and
```
if (something)
  return;
```
The former has behavior {Return}, the other has behavior {Nothing, Return}, since it is not guaranteed to return.

Reply:
Oh I see. So `Nothing` means that we can reach the end of the statement and go into the next one. I believe the name could reflect this semantics better. Nothing confused me :)

Reply:
I replaced Nothing by Default and added a sentence of explanation, please tell me if you can think of a clearer name.